docs/keyed_services.md

```diff
@@ -0,0 +1,184 @@
+              .Build()) { ... }
+```
+
+If you have more complicated logic then use
+`BrowserContextKeyedServiceFactory::GetBrowserContextToUse` and return nullptr
+if the service is not available for the given `BrowserContext`.
+
+```cpp
+content::BrowserContext* MyServiceFactory::GetBrowserContextToUse(
+    content::BrowserContext* context) const {
+  Profile* profile = Profile::FromBrowserContext(context);
+  return profile->IsRegularProfile() &&
+         base::FeatureList::IsEnabled(features::kMyFeature);
+}
+```
+
+Do not explicitly return null from other methods like
+`BuildServiceInstanceForBrowserContext`, `GetForProfile`, etc... If you return
+nullptr from `GetBrowserContextToUse` or configure `ProfileSelections`, the
+service will already be null. Avoid using helper methods like
+`IsAllowedForProfile` to see if a service is available for a particular profile
+as these are often used incorrectly/inconsistently. The preferred method is just
+a null check on the service itself. Prefer dependency injection
+[structure-modularity](https://chromium.googlesource.com/chromium/src/+/main/docs/chrome_browser_design_principles.md#structure_modularity)
+and pass in the services you require instead of calling the factory methods
+internally. This reduces dependencies on the factories and makes unit
+testing/mocking simpler.
+
+#### Keyed Services and Mojo
+
+Avoid adding wrapper classes around keyed services to implement mojo interfaces.
+It's usually better to implement them directly in the service. Factories should
+implement `GetRemoteForProfile` to return a `mojo::PendingRemote` and
+`BindRemoteForProfile` when you are passing in a `mojo::PendingReceiver`. The
+keyed service should implement `MakeRemote()` to return the
+`mojo::PendingRemote`. Please always ensure that you check to make sure the
+service is not null before calling `MakeRemote()` and return an empty pending
+remote if it is `mojo::PendingRemote<mojom::FakeService>()`.
+
+C++
+
+brave/components/fake/fake_service_impl.cc
+```cpp
+namespace fake {
+class FakeServiceImpl : public KeyedService, public mojom::FakeService {
+...
+}
+}  // namespace fake
+```
+
+brave/browser/fake/fake_service_factory.cc
+```cpp
+mojo::PendingRemote<mojom::FakeService> FakeServiceFactory::GetRemoteForProfile(
+    content::BrowserContext* context) {
+  auto* instance = GetInstance()->GetServiceForBrowserContext(context, true);
+  if (!instance) {
+    return mojo::PendingRemote<mojom::FakeService>();
+  }
+  return static_cast<fake::FakeServiceImpl*>(instance)->MakeRemote();
+}
+
+void FakeServiceFactory::BindRemoteForProfile(
+    content::BrowserContext* context,
+    mojo::PendingReceiver<fake::mojom::FakeService> receiver) {
+  auto* service = static_cast<fake::FakeServiceImpl*>(
+      GetInstance()->GetServiceForBrowserContext(context, true));
+  if (service) {
+    service->Bind(std::move(receiver));
+  }
+}
+
+static jlong JNI_FakeService_GetForProfile(
```

Comment:
Maybe similar to above comments from @goodov - having all the wording be `Example`. 
`JNI_ExampleService_` / `ExampleServiceFactory` / etc